conformance_test.go

```diff
@@ -0,0 +1,287 @@
+	if err != nil {
+		t.Fatal(err)
+	}
+	if sig.External != "" {
+		sigMsg.External = mustHexToBytes(sig.External)
+	}
+	err = sigMsg.Sign(new(zeroSource), signer)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = sigMsg.Verify(verifier)
+	if err != nil {
+		t.Fatal(err)
+	}
+	got, err := sigMsg.MarshalCBOR()
+	if err != nil {
+		t.Fatal(err)
+	}
+	want := mustHexToBytes(sig.Output.CBORHex)
+	if !deterministic {
+		got = got[:sig.OutputLength]
+		want = want[:sig.OutputLength]
+	}
+	if !bytes.Equal(want, got) {
+		t.Fatalf("unexpected output:\nwant: %x\n got: %x", want, got)
+	}
+}
+
+func getSigner(tc *TestCase, private bool) (cose.Signer, cose.Verifier, error) {
+	pkey, err := getKey(tc.Key, private)
+	if err != nil {
+		return nil, nil, err
+	}
+	alg := mustNameToAlg(tc.Alg)
+	signer, err := cose.NewSigner(alg, pkey)
+	if err != nil {
+		return nil, nil, err
+	}
+	verifier, err := cose.NewVerifier(alg, pkey.Public())
+	if err != nil {
+		return nil, nil, err
+	}
+	return signer, verifier, nil
+}
+
+func getKey(key Key, private bool) (crypto.Signer, error) {
+	switch key["kty"] {
+	case "RSA":
+		pkey := &rsa.PrivateKey{
+			PublicKey: rsa.PublicKey{
+				N: mustBase64ToBigInt(key["n"]),
+				E: mustBase64ToInt(key["e"]),
+			},
+		}
+		if private {
+			pkey.D = mustBase64ToBigInt(key["d"])
+			pkey.Primes = []*big.Int{mustBase64ToBigInt(key["p"]), mustBase64ToBigInt(key["q"])}
+			pkey.Precomputed = rsa.PrecomputedValues{
+				Dp:        mustBase64ToBigInt(key["dp"]),
+				Dq:        mustBase64ToBigInt(key["dq"]),
+				Qinv:      mustBase64ToBigInt(key["qi"]),
+				CRTValues: make([]rsa.CRTValue, 0),
+			}
+		}
+		return pkey, nil
+	case "EC":
+		var c elliptic.Curve
+		switch key["crv"] {
+		case "P-224":
+			c = elliptic.P224()
+		case "P-256":
+			c = elliptic.P256()
+		case "P-384":
+			c = elliptic.P384()
+		case "P-521":
+			c = elliptic.P521()
+		default:
+			return nil, errors.New("unsupported EC curve: " + key["crv"])
+		}
+		pkey := &ecdsa.PrivateKey{
+			PublicKey: ecdsa.PublicKey{
+				X:     mustBase64ToBigInt(key["x"]),
+				Y:     mustBase64ToBigInt(key["y"]),
+				Curve: c,
+			},
+		}
+		if private {
+			pkey.D = mustBase64ToBigInt(key["d"])
+		}
+		return pkey, nil
+	}
+	return nil, errors.New("unsupported key type: " + key["kty"])
+}
+
+// zeroSource is an io.Reader that returns an unlimited number of zero bytes.
+type zeroSource struct{}
+
+func (zeroSource) Read(b []byte) (n int, err error) {
+	for i := range b {
+		b[i] = 0
+	}
+
+	return len(b), nil
+}
+
+var encMode, _ = cbor.CanonicalEncOptions().EncMode()
+
+func decodeHeaders(protected, unprotected []byte) (hdr cose.Headers, err error) {
+	// test-vectors encodes the protected header as a map instead of a map wrapped in a bstr.
+	// UnmarshalFromRaw expects the former, so wrap the map here before passing it to UnmarshalFromRaw.
+	hdr.RawProtected, err = encMode.Marshal(protected)
+	if err != nil {
+		return
+	}
+	hdr.Protected = make(cose.ProtectedHeader)
```

Comment:
It is not necessary to do initialization. I think `UnmarshalFromRaw()` will handle it.